torchdynamo/guards.py

```diff
@@ -246,28 +350,28 @@ def CONSTANT_MATCH(self, guard: Guard):
 
     def NN_MODULE(self, guard: Guard):
         self.ID_MATCH(guard)
-        ref = self.arg_ref(guard)
-        val = self.get(guard.name)
-
-        def setup_guard():
-            assert istype(val.training, bool)
-            self.code.append(f"{ref}.training == {val.training}")
 
+        val = self.get(guard.name)
         if hasattr(val, "training"):
-            # There are cases where a monkeypatched object has a guard made between __new__ and __init__
-            setup_guard()
+            self.register_module_for_invalidation(val, self.guarded_code)
```

Comment:
I'd suggest registering this only on the first guard that this would remove.  It could be the case module has no guarded members and this isn't needed.

Reply:
I like this, let me try a lazy install here.

Reply:
I made some changes to make this possible, and it feels kinda gross, this feels cleaner, and if a module has no guarded members this invalidation is still pretty low overhead. Do you want me to commit up the alternative, lazy, approach to compare?